src/parser/Mediatype.h

```diff
@@ -0,0 +1,292 @@
+//
+//  parser/Mediatype.h
+//  librefract
+//
+//  Created by Jiri Kratochvil on 02/08/2019
+//  Copyright (c) 2019 Apiary Inc. All rights reserved.
+//
+
+#ifndef PARSER_MEDIATYPE_H
+#define PARSER_MEDIATYPE_H
+
+#include "../PEGTL/include/tao/pegtl.hpp"
+#include "../PEGTL/include/tao/pegtl/contrib/abnf.hpp"
+
+#include <string>
+#include <map>
+
+#include <iostream>
+
+namespace parser
+{
+
+    //
+    // Parsed for Media Types
+    // - type, subtype, suffix, paramemters attributes conforms to https://tools.ietf.org/html/rfc6838#section-4.2
+    // - parameter values conforms to https://tools.ietf.org/html/rfc2616#section-3.7
+    //
+
+    namespace mediatype
+    {
+
+        namespace pegtl = tao::pegtl;
+
+        // clang-format off
+
+        struct restricted : pegtl::one< '!', '#', '$', '&', '^',  '_', '-', '.' >  {};
+
+        using WS = pegtl::abnf::WSP; // White Space
+
+        using RWS = pegtl::plus<WS>; // Required White Space
+        using OWS = pegtl::star<WS>; // Optional White Space
+
+        using obs_text = pegtl::not_range< 0x00, 0x7F >;
+
+        struct quoted_pair : pegtl::if_must<pegtl::one<'\\'>, pegtl::sor<pegtl::abnf::VCHAR, obs_text, pegtl::abnf::WSP> > {};
+        // OPT: pegtl::if_must<> - if we want to throw on non-closed quotes
+        struct quoted : pegtl::seq<
+                            pegtl::one<'"'>,
+                            pegtl::star<pegtl::sor<quoted_pair, pegtl::not_one<'"'> >>,
+                            pegtl::one<'"'>
+                        > {};
+        struct name : pegtl::seq<
+                           pegtl::ascii::alnum,
+                           pegtl::rep_max<126, pegtl::sor<pegtl::ascii::alnum, restricted> >
+                       > {};
+        struct nonquoted : name {}; // nonqouted is used in parameter value to allow action template specialization
+
+        // parameters
+        struct attribute : name {};
+        struct value : pegtl::sor<quoted, nonquoted>  {};
+
+        // OPT convert to if_must<> to force correct attribute=value pair required
+        struct parameter : pegtl::seq<attribute, pegtl::one<'='>, value> {}; 
+        struct parameters : pegtl::list<parameter, RWS> {};
+
+        // type
+        struct type : name {};
+
+        // subtype 
+        struct subtype_nonprefixed : name {};
+        struct subtype_prefix : pegtl::plus<name, pegtl::one<'+'> > {};
+        struct subtype_suffix : name {};
+
+
+        struct subtype : pegtl::sor<
+                             pegtl::seq<subtype_prefix, subtype_suffix>,
+                             subtype_nonprefixed
+                         > {};
+
+        // separator
+        struct slash : pegtl::one<'/'> {};
+        // special case - if we do not want to throw on bad type definition like 'type/sub/type'
+        struct bad_slash : pegtl::not_at<slash> {};
```

Comment:
if user write something like `type/subtype/something` we are pretty sure the rest of line after second slash is defined wrong.
If we will remove `bad_slash` from `grammar` it will finish `parse` with `false` but `result` will be `{ "type", "subtype" }`. We will not know why.

This give us ability handle this specific error and also once parser will be completed, we can get user to know what is wrong.